### PR TITLE
feat(app): Integrate the block-sdk

### PR DESCRIPTION
**Pull request by g3ndo9kar9**
### In This PR 
- I integrate the block-sdk
    - I integrate the mev, default lanes (ordered accordingly) into the `LaneMempool` which will be used as the app's `Mempool`
    - I integrate the Prepare / ProcessProposal handlers and custom CheckTx handlers
         - Prepare / ProcessProposal will be responsible for building blocks
         - The CheckTx handler is responsible for
             - Performing extra verification on bid-txs
             - Purging txs from cometBFT's mempool if they aren't in the app
    - Modules
        - I integrate the `x/auction` module, and the corresponding `AnteHandler` for the auction module
### Testing
- We've tested this integration against our own set of load-tests. These tests, spin up a network of 5 nodes (3 val, 1 seed, 1 full). Perform various perturbations against them (killing, restarting, disconnecting from the network, etc.), and periodically send a load of 350 tx/sec for 90s. The output from one of our trials is [here](https://snapshots.raintank.io/dashboard/snapshot/y23ErIIGdQEL8AXCbFdv4grb2hxldbMz?orgId=2), happy to go through these in more depth / review the methodology.
### Let's Discuss
- I configured the x/auction module in the `v202` upgrade handler. The params chosen are [here](https://github.com/neutron-org/neutron/pull/413/files#diff-7c95504b0c554fdd10576f31ea32ca63a0321c2ddf0bb85f79821c3257b0b338R31), happy to change these if requested
- I also configured the lane mempool capacities / other parameters [here](https://github.com/neutron-org/neutron/pull/413/files#diff-802c8455e6cfe3da0a5fb5d9ff1f32ca2e651f0a0434163eb700887a0c871391R20)

### CODE REVIEW

**Comment by pr0n00gler** [comment]
Thx!

File: app/upgrades/v2.0.2/constants.go [hunk 1]
```diff
@@ -1,15 +1,32 @@
...
-}
+var (
+	Upgrade = upgrades.Upgrade{
+		UpgradeName:          UpgradeName,
+		CreateUpgradeHandler: CreateUpgradeHandler,
+		StoreUpgrades: storetypes.StoreUpgrades{
+			Added: []string{auctiontypes.ModuleName},
+		},
+	}
+	AuctionParamsReserveFee      = sdk.Coin{Denom: "untrn", Amount: sdk.NewInt(1_000_000)}
```
Comment (mmikev) [inline app/upgrades/v2.0.2/constants.go]: Please make it 500_000 (0.5NTRN)

File: app/upgrades/v2.0.2/constants.go [hunk 2]
```diff
@@ -1,15 +1,32 @@
...
+	AuctionParamsMinBidIncrement = sdk.Coin{Denom: "untrn", Amount: sdk.NewInt(1_000_000)}
```
Comment (mmikev) [inline app/upgrades/v2.0.2/constants.go]: Please make it 100_000 (0.1NTRN)

File: app/upgrades/v2.0.2/upgrades.go [hunk 3]
```diff
@@ -1,25 +1,51 @@
...
+		err := setAuctionParams(ctx, keepers.FeeBurnerKeeper, keepers.AuctionKeeper)
+		if err != nil {
+			return nil, err
+		}
+
 		return vm, nil
 	}
 }
+
+func setAuctionParams(ctx sdk.Context, feeBurnerKeeper *feeburnerkeeper.Keeper, auctionKeeper auctionkeeper.Keeper) error {
```
Comment (mmikev) [inline app/upgrades/v2.0.2/upgrades.go]: Could you create a `nextupgrade` folder in the `upgrades` folder and put this params set in the newly created upgrade handler, pls?

Like here: https://github.com/neutron-org/neutron/pull/414/files#diff-16b23f75ba28a93ca5912d821b29a53db322c7736f895454870c6e318efdecfe

**Review by pr0n00gler — CHANGES_REQUESTED** [review]
_here was a draft comment, removed :)_

File: app/upgrades/v2.0.2/upgrades.go (hunk 3 above)
Comment (mmikev) [reply app/upgrades/v2.0.2/upgrades.go]: Basically move all block-sdk related stuff into the `nextupgrade` folder instead of `2.0.2`, please

**Review by pr0n00gler — CHANGES_REQUESTED** [review]
[Tests](https://github.com/neutron-org/neutron/actions/runs/7631527306/job/20789666853?pr=413) do no pass unfortunately :(
You can run it via `make test`

Please fix it:

```
--- FAIL: TestKeeperTestSuite (0.19s)
    --- FAIL: TestKeeperTestSuite/TestAuctionUpgrade (0.19s)
        suite.go:87: test panicked: ApplyUpgrade should never be called without first checking HasHandler
            goroutine 24 [running]:
            runtime/debug.Stack()
            	/usr/local/go/src/runtime/debug/stack.go:24 +0x64
            github.com/stretchr/testify/suite.failOnPanic(0x14000f084e0, {0x103212ba0, 0x1036a76f0})
...
```

File: app/app.go [hunk 4]
```diff
@@ -928,6 +968,19 @@ func New(
 	app.SetInitChainer(app.InitChainer)
 	app.SetBeginBlocker(app.BeginBlocker)
 
+	app.SetEndBlocker(app.EndBlocker)
```
Comment (swelf19) [inline app/app.go:966]: the same method called on L1016, pls remove one of

File: app/app.go [hunk 5]
```diff
@@ -942,17 +995,56 @@ func New(
...
 	app.SetAnteHandler(anteHandler)
 
+	// set ante-handlers
+	opts := []base.LaneOption{
+		base.WithAnteHandler(anteHandler),
+	}
+	baseLane.WithOptions(opts...)
+	mevLane.WithOptions(opts...)
+
 	app.SetEndBlocker(app.EndBlocker)
```
Comment (swelf19) [inline app/app.go]:
```suggestion
```

File: app/app.go [hunk 6]
```diff
@@ -324,6 +342,14 @@ type App struct {
...
 	// sm is the simulation manager
 	sm *module.SimulationManager
+
+	// Custom checkTx handler -> this check-tx is used to simulate txs that are
+	// wrapped in a bid-tx
+	checkTxHandler checktx.CheckTx
+
+	// MEVLane is used as a dependency for the x/auction module's antehandler. This references th lane in the app's
+	// LaneMempool.
+	MEVLane auctionante.MEVLane
```
Comment (swelf19) [inline app/app.go]: Not sure we need to define a field for `auctionante.MEVLane`, the only place you use it is L1000, passing it to antehandler, but you can do it without an app's field

File: go.mod [hunk 7]
```diff
@@ -1,6 +1,6 @@
 module github.com/neutron-org/neutron/v2
 
-go 1.21.1
+go 1.21.4
```
Comment (g3ndo9kar9) [inline go.mod:3]: patch version updated so that test / lint would run successfully

File: app/app.go [hunk 8]
```diff
@@ -942,16 +990,57 @@ func New(
...
+		app.GetTxConfig().TxDecoder(),
+		app.GetTxConfig().TxEncoder(),
+		mempool,
+	)
+	app.SetPrepareProposal(handler.PrepareProposalHandler())
+
+	// we use a no-op ProcessProposal, this way, we accept all proposals in avoidance
+	// of liveness failures due to Prepare / Process inconsistency. In other words,
+	// this ProcessProposal always returns ACCEPT.
+	app.SetProcessProposal(baseapp.NoOpProcessProposal())
```
Comment (g3ndo9kar9) [inline app/app.go:1023]: Setting ProcessProposal to a nop

**Review by pr0n00gler — APPROVED** [review]

**Pull request merged**